src/plugins/auto/plugin.cpp

```diff
@@ -406,96 +423,50 @@ DeviceInformation MultiDeviceInferencePlugin::SelectDevice(const std::vector<Dev
             continue;
         }
     }
 
-    if (CPU.empty() && dGPU.empty() && iGPU.empty() && MYRIAD.empty() && VPUX.empty()) {
-        IE_THROW(NotFound) << "No available device found";
-    }
-
     // Priority of selecting device: dGPU > VPUX > iGPU > MYRIAD > CPU
-    if (!dGPU.empty()) {
-        for (auto&& item : dGPU) {
-            std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-            auto supportNetwork = std::find(capability.begin(), capability.end(), networkPrecision);
-            if (supportNetwork != capability.end()) {
-                return item;
-            }
-        }
-    } else if (!VPUX.empty()) {
-        for (auto&& item : VPUX) {
-            std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-            auto supportNetwork = std::find(capability.begin(), capability.end(), networkPrecision);
-            if (supportNetwork != capability.end()) {
-                return item;
-            }
-        }
-    } else if (!iGPU.empty()) {
-        for (auto&& item : iGPU) {
-            std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-            auto supportNetwork = std::find(capability.begin(), capability.end(), networkPrecision);
+    std::list<DeviceInformation>  devices;
+    devices.splice(devices.end(), dGPU);
+    devices.splice(devices.end(), VPUX);
+    devices.splice(devices.end(), iGPU);
+    devices.splice(devices.end(), MYRIAD);
+
+    std::list<DeviceInformation> validDevices;
+
+    auto selectSupportDev = [this, &devices, &validDevices](const std::string& networkPrecision) {
+        for (auto iter = devices.begin(); iter != devices.end();) {
+            std::vector<std::string> capability = GetCore()->GetMetric(iter->deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+            auto supportNetwork = std::find(capability.begin(), capability.end(), (networkPrecision));
             if (supportNetwork != capability.end()) {
-                return item;
-            }
-        }
-    } else if (!MYRIAD.empty()) {
-        for (auto&& item : MYRIAD) {
-            std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-            auto supportNetwork = std::find(capability.begin(), capability.end(), networkPrecision);
-            if (supportNetwork != capability.end()) {
-                return item;
+                validDevices.push_back(std::move(*iter));
+                devices.erase(iter++);
+                continue;
             }
+            iter++;
         }
-    }
-
-    // If network is FP32 but there is no device support FP32, offload FP32 network to device support FP16.
+    };
+    selectSupportDev(networkPrecision);
+    // If network is FP32, continue to collect the device support FP16 but not support FP32.
```

Comment:
If the network is FP32 **but there is no FP32 device in the system**, continue to collect the device support FP16 but not support FP32.